File: Programmers/[72414] 광고 삽입/JFe/72414.py
def solution(play_time, adv_time, logs):
    tmp = play_time.split(':')
    play_time = int(tmp[0]) * 3600 + int(tmp[1]) * 60 + int(tmp[2])
    time_list = [0 for _ in range(play_time + 1)]
    # log 시간 변환
    for log in logs:
        l = log.split('-')
        for i, v in enumerate(l):
            tmp = v.split(':')
            l[i] = int(tmp[0]) * 3600 + int(tmp[1]) * 60 + int(tmp[2])
        time_list[l[0]] += 1
        time_list[l[1]] -= 1
    # 구간별 시청자 수 저장
    for i in range(1, len(time_list)):
        time_list[i] += time_list[i-1]
    # 누적 시청자 수 저장
    for i in range(1, len(time_list)):
        time_list[i] += time_list[i-1]
    # 광고 시간 변환
    tmp = adv_time.split(':')
    adv_time = int(tmp[0]) * 3600 + int(tmp[1]) * 60 + int(tmp[2])
    # 시청자 수 가장 많은 구간 체크
    most_view = time_list[adv_time-1]
    result = 0
    for i in range(adv_time, play_time):
        if most_view < time_list[i] - time_list[i - adv_time]:
            most_view = time_list[i] - time_list[i - adv_time]
            result = i - adv_time + 1
    # 결과 값 문자열 변환
    hour = '0' + str(result // 3600)
    min = '0' + str(result % 3600 // 60)
    sec = '0' + str(result % 60)
    return hour[-2:] + ':' + min[-2:] + ':' + sec[-2:]

File: Programmers/[72414] 광고 삽입/KingDonggyu/72414.js
function solution(play_time, adv_time, logs) {
  play_time = getSeconds(play_time);
  adv_time = getSeconds(adv_time);

  const playCount = Array(play_time + 1).fill(0);

  logs.map((log) => {
    const [start, end] = log.split('-');
    playCount[getSeconds(start)]++;
    playCount[getSeconds(end)]--;
  });

  // 시청자 수 구하기
  for (let i = 1; i <= play_time; i++) {
    playCount[i] += playCount[i - 1];
  }

  // 누적 시청자 수 구하기
  for (let i = 1; i <= play_time; i++) {
    playCount[i] += playCount[i - 1];
  }

  let answer = 0;
  let cumulativeTime = playCount[adv_time - 1];

  // 누적 재생시간 최대값 구하기
  for (let i = adv_time; i <= play_time; i++) {
    if (cumulativeTime < playCount[i] - playCount[i - adv_time]) {
      cumulativeTime = playCount[i] - playCount[i - adv_time];
      answer = i - adv_time + 1;
    }
  }

  return formatTime(answer).join(':');
}

const getSeconds = (time) => {
  const [h, m, s] = time.split(':');
  return h * 3600 + m * 60 + s * 1;
};

const formatTime = (seconds) => {
  let h = Math.floor(seconds / 3600);
  let m = Math.floor((seconds - hour * 3600) / 60);
  let s = seconds - hour * 3600 - minute * 60;

  return [h, m, s].map((time) => (time > 9 ? time : '0' + time));
};
